docs/designs/ps_design.md

```diff
@@ -1,377 +1,146 @@
+1. model sharding
+2. high performance
+3. fault-tolerance
 
-## PS
+## Model Sharding
+There are two kinds of parameters in the training process, embedding parameters and non-embedding parameters.
 
-### Compoments
+Embedding parameters consist of multiple embedding tables. Each embedding table correponds to one embedding layer in the user-defined model structure. An embedding table is a data structure that maps a discrete value, named embedding id *id*, to a vector, named embedding vector *vector*. Since embedding parameters might be up to terabytes, we can distribute them by placing different embedding vectors on different parameter server instance. 
 
-![pserver](../images/pserver.png)
+Non-embedding parameters are in the form multiple dense tensors. Theoretically, dense tensors might have tremendous size and require sharding. However, in practices, researchers don't often define models depending on huge dense tensor parameters. Hence in this design, we don't partition dense tensors; instead, we place each dense tensor on a parameter server instance.
 
-PS contains two main compoments:
+For each dense tensor or an embedding vector, denoted by x, we put it on the parameter server p(x). For a dense tensor x, we denote key(x) for its name; for an embedding vector x, key(x) consists of the name of the embedding layer and its embedding id.
 
-- KVStore
-- Optimizer
+There are many ways to define p(x). For simplicity, we choose a fixed mapping from x to a range of PS instances [0, N):
 
-The worker initializes a model, and pushes parameters to KVStore. Before each step of training, the worker pulls the latest model from KVStore. After a round of forward/backward computation, the worker pushes gradients to the PS waiting for processing. Then, optimizer of PS will look up the corresponding parameter from KVStore. At last, it applies gradients to parameters, and updates parameter back to KVStore.
+p(x) = hash(key(x)) % N
 
+It is noticeable that Kubernetes might preempt some parameter server instances. In such a case, we might be afraid that N isn't constant. However, we can overcome this case by setting parameter server instances having higher priority than worker processes in a job. By doing so, preemption kills workers other than parameter servers. If all workers are dead, the job stops until there comes free resources, and Kubernetes starts some workers for the job.  For more information about preemption, please refer to [this document](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/). In short, we can assume that N is a constant number in the Kubernetes-native architecture of ElasticDL.
```

Comment:
In k8s, we use pod concept.
parameter server instance --> parameter server pod

Reply:
In my mind, instance and pod are both ok in this design doc. But instance may be more friendly to people who are familiar with machine learning, but not familiar with k8s. Thus I prefer instance.

Reply:
I think we can use instances or processes here.